File: app/game.lua
function field_cell_move(i, ii, x, y)
    if i + x == 0 or i + x == 4 then
        return
    end
    if field[i+x][ii+y] == 0 then
        field[i+x][ii+y] = field[i][ii]
        field[i][ii] = 0
        field_cell_move(i+x, ii+y, x, y)
        -- если ячейки одинаковые - объеденить
    elseif field[i+x][ii+y] == field[i][ii] then
        field[i+x][ii+y] = field[i][ii]*2
        field[i][ii] = 0
    end
end
function field_add_new()
    local freeCell = false
    do
        local i = 1
        while i <= 3 do
            do
                local ii = 1
                while ii <= 4 do
                    if field[i][ii] == 0 then
                        field[i][ii] = 2
                        freeCell = true
                        return
                    end
                    ii = ii + 1
                end
            end
            i = i + 1
        end
    end
    if not freeCell then
        restart()
    end
end
function restart()
    field = {{0, 2, 0, 2}, {2, 2, 0, 0}, {0, 0, 0, 0}}
    update_ui_field()
end
function update_ui_field()
    do
        local i = 1
        while i <= 3 do
            do
                local ii = 1
                while ii <= 4 do
                    local name = table.concat({i-1,ii-1})
                    gui.set_text(gui.get_node(name), field[i][ii])
                    ii = ii + 1
                end
            end
            i = i + 1
        end
    end
end
field = {{0, 2, 0, 2}, {2, 2, 0, 0}, {0, 0, 0, 0}}
function field_move(x, y)
    do
        local i = 1
        while i <= 3 do
            do
                local ii = 1
                while ii <= 4 do
                    if x > 0 then
                        field_cell_move(4 - i, ii, x, y)
                    elseif x < 0 then
                        field_cell_move(i, ii, x, y)
                    elseif y > 0 then
                        field_cell_move(i, 5 - ii, x, y)
                    else
                        field_cell_move(i, ii, x, y)
                    end
                    ii = ii + 1
                end
            end
            i = i + 1
        end
    end
    field_add_new()
    update_ui_field()
end

File: src/game.ts
let field: number[][] = [[0, 2, 0, 2], [2, 2, 0, 0], [0, 0, 0, 0]];

// Перемещает ячейки поля по 4 сторонам
function field_move(x: number, y: number): void {

  for (let i = 1; i <= 3; i++) {
    for (let ii = 1; ii <= 4; ii++) {
      if (x > 0) {
        field_cell_move(4 - i, ii, x, y);
      } else if (x < 0) {
        field_cell_move(i, ii, x, y);
      } else if (y > 0) {
        field_cell_move(i, 5 - ii, x, y);
      } else {
        field_cell_move(i, ii, x, y);
      }
    }
  }
  field_add_new();
  update_ui_field();
}

// Перемещает ячейку поля, вызывается из field_move
function field_cell_move(i: number, ii: number, x: number, y: number): void {

  // если ячейка пустая - переместить
  if (i + x === 0 || i + x === 4) return;
  if (field[i + x][ii + y] === 0) {
    field[i + x][ii + y] = field[i][ii];
    field[i][ii] = 0;
    field_cell_move(i + x, ii + y, x, y);
    // если ячейки одинаковые - объединить
  } else if (field[i + x][ii + y] === field[i][ii]) {
    field[i + x][ii + y] = field[i][ii] * 2;
    field[i][ii] = 0;
  }
}

// Случайно выбирает ячейку для спавна новой 2
function field_add_new(): void {

  let freeCell = false;
  for (let i = 1; i <= 3; i++) {
    for (let ii = 1; ii <= 4; ii++) {
      if (field[i][ii] === 0) {
        field[i][ii] = 2;
        freeCell = true;
        return;
      }
    }
  }
  if (!freeCell) {
    restart();
  }
}

// Перезапускает field если field_add_new не найдёт пустую ячейку
function restart(): void {

  field = [[0, 2, 0, 2], [2, 2, 0, 0], [0, 0, 0, 0]];
  update_ui_field();
}

// Обновляет UI после field_move и field_add_new
function update_ui_field(): void {

  for (let i = 1; i <= 3; i++) {
    for (let ii = 1; ii <= 4; ii++) {
      const name = `${i - 1}${ii - 1}`;
      gui.set_text(gui.get_node(name), field[i][ii]);
      // gui.set_color(gui.get_node("box"), color);
    }
  }
}
